template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -243,7 +250,50 @@ private Map<String, Set<ServiceComponent>> getNonGatewayServicesByHostGroup() {
                                 .filter(i -> !i.getComponent().equalsIgnoreCase("GATEWAY"))
                                 .collect(toSet())
                 ));
+    }
+
+    private Set<String> collectComponentByHostGroupWithRoleDependencies() {
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        Set<String> services = hgToNonGwServiceComponents.values().stream()
+                .filter(serviceComponents -> !Strings.isNullOrEmpty(getDependentService(serviceComponents)))
+                .map(serviceComponents -> ROLE_DEPENDENCY.get(getDependentService(serviceComponents)))
+                .collect(toSet());
+        return services;
+    }
+
+    private String getDependentService(Set<ServiceComponent> serviceComponents) {
+        Set<String> serviceComponentRoles = collectComponents(serviceComponents);
+        return ROLE_DEPENDENCY.keySet().stream().filter(serviceComponentRoles::contains).findFirst().orElse(null);
+    }
+
+    public List<String> getUnhealthyDependentHosts(StackDto stackdto) {
+        Map<String, Set<ServiceComponent>> nonGatewayServicesByHostGroup = getNonGatewayServicesByHostGroup();
+        Map<String, Set<String>> nonGatewayServices = nonGatewayServicesByHostGroup.entrySet().stream()
+                .collect(toMap(
+                        Entry::getKey,
+                        e -> e.getValue().stream()
+                                .map(ServiceComponent::getComponent)
+                                .collect(toSet())
+                ));
+
+        Set<String> dependentComponent = collectComponentByHostGroupWithRoleDependencies();
+        Set<String> dependentHostNames = nonGatewayServices.entrySet().stream()
+                .filter(component -> isComponentPresentInHost(component.getValue(), dependentComponent))
+                .map(Entry::getKey)
+                .collect(toSet());
+        List<InstanceMetadataView> unhealthyInstanceMetadata = stackdto.getAllAvailableInstances().stream()
+                .filter(instanceMetaData -> InstanceStatus.SERVICES_UNHEALTHY.equals(instanceMetaData.getInstanceStatus()))
+                .collect(toList());
+        List<String> allunHealthyHostsNames = unhealthyInstanceMetadata.stream()
+                .map(instanceMetadata -> instanceMetadata.getInstanceGroupName()).collect(toList());
+        List<String> unhealthyHostsNames = dependentHostNames.stream()
+                .filter(i -> allunHealthyHostsNames.contains(i))
+                .collect(toList());
+        return unhealthyHostsNames;
```

Comment:
The nomenclature here is a bit confusing. `hostNames` is different from instanceGroupName / hostGroupName which is being collected into these lists.

Reply:
Okay will change it

Reply:
Done